### PR TITLE
Filter Envoy Access Logs by status code

### PR DESCRIPTION
**Pull request by andrzejwaw**
Allows filtering access logs by status code. 
Pass metadata ```access_log_filter_http_code``` variable with operator and status code to filter. Variable expected format: ```{operator}:{status code}```. Allowed operators are:
* le - lower equal
* eq - equal
* ge - greater equal

### CODE REVIEW

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt [hunk 1]
```diff
@@ -64,11 +69,38 @@ class MetadataNodeGroup(val properties: SnapshotProperties) : NodeGroup<Group> {
         return ListenersHostPortConfig(ingressHost, ingressPort, egressHost, egressPort)
     }
 
+    private fun metadataToAccessLogFilterHttpCode(metadata: Struct): AccessLogFilterHttpCode? {
+        val httpFilterCode = metadata.fieldsMap["access_log_filter_http_code"]?.stringValue
```
Comment (lukidzi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt:73]: I would change a structure of metadata:
```
access_log_filter:
  status_code_filter: 
    status_code:
    operator:
```
So in future we will have elastic structure which allows to add new filters(header, flags, ..)

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt [hunk 2]
```diff
@@ -64,11 +69,38 @@ class MetadataNodeGroup(val properties: SnapshotProperties) : NodeGroup<Group> {
...
+        var op: Op? = null
+        var code: String? = null
+
+        if (!httpFilterCode.isNullOrBlank()) {
+            val regex = """(eq|ge|le?):(\d+)""".toRegex()
```
Comment (chemicL) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt:78]: we can just use split by colon and eliminate the need for regex

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt (hunk 1 above)
Comment (chemicL) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt:73]: As discussed today, let's go with this:
```
access_log_filter:
  status_code_filter: "ge:400"
```

We get raw string, colon separated. We split by colon and extract `status_code` and `operator` from that.
In the future if required we can add `header_filter` or extend `status_code_filter` by supporting comma-separated values.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt (hunk 2 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt]: I can change it to simple split, but personally I prefer keep it. This is simple regex, even made it more simpler and strict in current PR. I think having regex allows us use less ifs to check pattern.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 3]
```diff
@@ -54,6 +74,33 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
     }
 }
 
+fun Value?.toStatusCodeFilter(properties: AccessLogFilterProperties): AccessLogFilter.StatusCodeFilter? {
+    val value = this?.stringValue
+
+    if (value != null) {
+        var op: ComparisonFilter.Op? = null
+        val regex = """((le)|(eq)|(ge)){1}:(\d{3})""".toRegex()
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:82]: Won't `toRegex()` make issues here, like a recent memory pressure? I think we should consistently use re2j with a static matcher so we won't make accidentally the same mistake again.

Comment (pzmi) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:82]: I think @slonka fixed that issue.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 4]
```diff
@@ -54,6 +74,33 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
...
+                "eq" -> op = ComparisonFilter.Op.EQ
+                "ge" -> op = ComparisonFilter.Op.GE
+            }
+            return AccessLogFilter.StatusCodeFilter(
+                comparisonOP = op!!,
+                comparisonCode = matchResult.groupValues.get(properties.codeIndex).toInt()
+            )
+        } ?: run {
+            throw NodeMetadataValidationException(
+                "Access log filter status code doe not match pattern: ((le)|(eq)|(ge)){1}:(\\d{3})"
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:97]: Maybe it's my personal reluctance towards regexes but I'd rather have more straightforward like
`Invalid access log status code filter. Expected OPERATION:STATUS_CODE`

Also typo in "does"

**Comment by pzmi** [comment]
Shouldn't we add some documentation for new metadata field?

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 5]
```diff
@@ -21,6 +23,24 @@ class NodeMetadata(metadata: Struct, properties: SnapshotProperties) {
     val communicationMode = getCommunicationMode(metadata.fieldsMap["ads"])
 
     val proxySettings: ProxySettings = ProxySettings(metadata.fieldsMap["proxy_settings"], properties)
+
+    val accessLogFilter: AccessLogFilter = AccessLogFilter(
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:27]: It looks like it isn't used anywhere or my IDE just isn't seeing any usage.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 3 above)
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:82]: `{1}` is superfluous

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 6]
```diff
@@ -54,6 +74,33 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
...
 
+fun Value?.toStatusCodeFilter(properties: AccessLogFilterProperties): AccessLogFilter.StatusCodeFilter? {
+    val value = this?.stringValue
+
+    if (value != null) {
+        var op: ComparisonFilter.Op? = null
+        val regex = """((le)|(eq)|(ge)){1}:(\d{3})""".toRegex()
+
+        val matchResult = regex.matchEntire(value.toLowerCase())
+        matchResult?.takeIf { !it.groupValues.isEmpty() }?.apply {
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:85]:
```suggestion
        matchResult?.takeIf { it.groupValues.isNotEmpty() }?.apply {
```

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 7]
```diff
@@ -54,6 +74,33 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
...
+            }
+            return AccessLogFilter.StatusCodeFilter(
+                comparisonOP = op!!,
+                comparisonCode = matchResult.groupValues.get(properties.codeIndex).toInt()
+            )
+        } ?: run {
+            throw NodeMetadataValidationException(
+                "Access log filter status code doe not match pattern: ((le)|(eq)|(ge)){1}:(\\d{3})"
+            )
+        }
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:99]: You don't need `run` here.
```suggestion
        } ?: throw NodeMetadataValidationException(
            "Access log filter status code doe not match pattern: ((le)|(eq)|(ge)){1}:(\\d{3})"
        )
```

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 8]
```diff
@@ -54,6 +74,33 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
...
+fun Value?.toStatusCodeFilter(properties: AccessLogFilterProperties): AccessLogFilter.StatusCodeFilter? {
+    val value = this?.stringValue
+
+    if (value != null) {
+        var op: ComparisonFilter.Op? = null
+        val regex = """((le)|(eq)|(ge)){1}:(\d{3})""".toRegex()
+
+        val matchResult = regex.matchEntire(value.toLowerCase())
+        matchResult?.takeIf { !it.groupValues.isEmpty() }?.apply {
+            when (matchResult.groupValues.get(properties.operatorIndex)) {
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:86]: You can move `op` declaration here, so you don't need unsafe access later.
```kotlin
            val op = when (matchResult.groupValues[properties.operatorIndex]) {
                "le" -> ComparisonFilter.Op.LE
                "eq" -> ComparisonFilter.Op.EQ
                "ge" -> ComparisonFilter.Op.GE
                else -> throw NodeMetadataValidationException("Invalid operator. Allowed values: le,eq,ge")
            }
            return AccessLogFilter.StatusCodeFilter(
                comparisonOP = op,
```

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt [hunk 9]
```diff
@@ -24,6 +24,12 @@ class SnapshotProperties {
     var enabledCommunicationModes = EnabledCommunicationModes()
     var shouldSendMissingEndpoints = false
     var metrics: MetricsProperties = MetricsProperties()
+    var accessLogFilterProperties: AccessLogFilterProperties = AccessLogFilterProperties()
+}
+
+class AccessLogFilterProperties {
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt:30]: Why do we need these indexes? Can filter regex be modified somehow?

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 3 above)
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:82]: Why group each operator? `val regex = """(le|eq|ge):(\d{3})""".toRegex()` This should be enough. It also simplifies matching groups. I was wondering why the status code is group 5 instead of 2. Without these additional groups, it will be 2 :)

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt [hunk 10]
```diff
@@ -313,10 +317,36 @@ class EnvoyListenersFactory(
...
+                                            .setRuntimeKey("access_log_filter_http_code")
+                                            .build()
+                                    )
+                                    .build()
+                            )
+                            .build()
+                    )
+                )
+            }
+        }
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt:347]: This is some christmas tree code :D How about extracting an innermost builder?
```suggestion
        accessLogFilterHttpCode?.let {
            it.statusCodeFilter?.let { statusCodeFilter ->
                builder.setFilter(accessLogFilterFrom(statusCodeFilter))
            }
        }
```

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt [hunk 11]
```diff
@@ -313,10 +317,36 @@ class EnvoyListenersFactory(
...
     }
 
-    private fun accessLog(accessLogPath: String, accessLogType: String): AccessLog {
-        return AccessLog.newBuilder()
-                .setName("envoy.file_access_log")
-                .setTypedConfig(
+    private fun accessLog(
+        accessLogPath: String,
+        accessLogType: String,
+        accessLogFilterHttpCode: pl.allegro.tech.servicemesh.envoycontrol.groups.AccessLogFilter?
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt:323]: What do you think about aliasing this or Enovy's one?
```kotiln
import io.envoyproxy.envoy.config.filter.accesslog.v2.AccessLogFilter as EnvoyAccessLogFilter
```

Comment (pzmi) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt:323]: Or maybe changing its name to `AccessLogFilterSettings` as it isn't a filter but just settings. Then I'd change the parameter name, `accessLogFilterHttpCode`, to something more adequate.

Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt:323]: will change it to AccessLogFilterSettings

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt (hunk 9 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt:30]: right, we do not need them

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 3 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:82]: I'll use re2j pattern

File: docs/features/access_log_filter.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+Configuration provided in `metadata.access_log_filter` will be used to set up an access log filter for `Envoy`.
+
+## Filter logs by status code
+
+Envoy allows filtering access logs by status code.
+Variable `metadata.access_log_filter.status_code_filter` should contain operator and status code.
+
+Expected format is: `{operator}`:`{status code}`.
+
+Allowed operators are:
```
Comment (lukidzi) [inline docs/features/access_log_filter.md:14]: We should add information if operator should be lower/upper case or it is case insensitive

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 13]
```diff
@@ -54,6 +69,27 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
...
+    val split = value.split(properties.delimiter)
+    val op = when (split[0]) {
+        "le" -> ComparisonFilter.Op.LE
+        "eq" -> ComparisonFilter.Op.EQ
+        "ge" -> ComparisonFilter.Op.GE
+        else -> throw NodeMetadataValidationException("Invalid operator. Allowed values: le,eq,ge")
+    }
+    return AccessLogFilterSettings.StatusCodeFilter(
+        comparisonOP = op,
+        comparisonCode = split[1].toInt()
```
Comment (lukidzi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:89]: This line might throw null pointer if split return a list with 1 element.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 14]
```diff
@@ -54,6 +69,27 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
     }
 }
 
+fun Value?.toStatusCodeFilter(properties: AccessLogFilterProperties): AccessLogFilterSettings.StatusCodeFilter? {
+    val value = if (this?.stringValue != null) this.stringValue.toLowerCase() else return null
+
+    if (!properties.statusCodeFilterPattern.matches(value)) {
```
Comment (lukidzi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:75]: I am not sure if `statusCodeFilterPattern` property is a good idea. If we change property I expect it to work but few lines below we have hardcoded values

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt [hunk 15]
```diff
@@ -313,10 +319,20 @@ class EnvoyListenersFactory(
...
+    private fun accessLog(
+        accessLogPath: String,
+        accessLogType: String,
+        accessLogFilterSettings: AccessLogFilterSettings?
+    ): AccessLog {
+        val builder = AccessLog.newBuilder().setName("envoy.file_access_log")
+
+        accessLogFilterSettings?.let { settings ->
+            settings.statusCodeFilter?.let {
+                builder.setFilter(accessLogFilterValue(it.comparisonOP, it.comparisonCode))
```
Comment (lukidzi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt:331]: `comparisonOP` -> `comparisonOp`

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 13 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:89]: If checked above pattern will match, split would not return 1 element, regardless of that should I catch such case?

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt (hunk 15 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt]: changed to `comparisonOperator`

File: docs/features/access_log_filter.md (hunk 12 above)
Comment (andrzejwaw) [reply docs/features/access_log_filter.md:14]: added in new commit

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt (hunk 10 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt]: done

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 13 above)
Comment (lukidzi) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:89]: What if I will change pattern and it won't return 2 element list? Also pattern property IMO should be configurable to any pattern I want but know it support only `eq|le|ge:d+{3}` so maybe it shouldn't be configurable ?

Comment (slonka) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:89]: I would not make pattern configurable since it is tightly bound to the code here.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 14 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:75]: will remove those hardcoded values

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 13 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:83]: in new commit I removed hardcoded condition. Please refer, is it better right now?

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogStatusCodeFilterFactory.kt [hunk 16]
```diff
@@ -0,0 +1,27 @@
+package pl.allegro.tech.servicemesh.envoycontrol.groups
+
+import com.google.re2j.Pattern
+import io.envoyproxy.envoy.config.filter.accesslog.v2.ComparisonFilter
+
+class AccessLogStatusCodeFilterFactory {
```
Comment (lukidzi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogStatusCodeFilterFactory.kt:6]: Let's call this class `AccessLogFilterFactory` and move logic responsible for access log configuration from `EnvoyListenerFactory`.

Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogStatusCodeFilterFactory.kt]: Class renamed and moved building status code filter to factory

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt (hunk 1 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt]: done

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+import io.envoyproxy.envoy.config.filter.accesslog.v2.AccessLog.Builder
+import io.envoyproxy.envoy.config.filter.accesslog.v2.StatusCodeFilter.newBuilder
+
+class AccessLogFilterFactory {
+    private val operators: Array<ComparisonFilter.Op> = arrayOf(
+        ComparisonFilter.Op.LE, ComparisonFilter.Op.GE, ComparisonFilter.Op.EQ
+    )
+    private val delimiter: Char = ':'
+    private val statusCodeFilterPattern: Pattern = Pattern.compile(
+        """(${operators.joinToString("|")})$delimiter(\d{3})"""
```
Comment (lukidzi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt:16]: one more this. Pattern should looks like `^(${operators.joinToString("|")})$delimiter(\d{3})$` because now `testeq:400test` works

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt [hunk 18]
```diff
@@ -318,4 +333,42 @@ class NodeMetadataTest {
...
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        "LT:123",
+        "equal:400",
+        "eq:24",
+        "GT:200"
+    )
+    fun `should not set statusCodeFilter for accessLogFilter`(input: String) {
```
Comment (slonka) [inline envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt:360]: In the previous test there is MethodSource: `validStatusCodeFilterData` -> that way I know why it should set statusCodeFilter, in this method you used @CvsSource and I do not know why it should not be set. I would clarify in the test name (something like: `should not set statusCodeFilter for accessLogFilter for invalid status code filter data`) or use MethodSource like before.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt (hunk 17 above)
Comment (slonka) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt:16]: Please change this.

Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt:16]: Changed, but can not agree that `testeq:400test`was matching. Pattern is validated for whole input https://github.com/google/re2j/blob/re2j-1.3/java/com/google/re2j/Pattern.java#L144

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt (hunk 18 above)
Comment (andrzejwaw) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt]: changed test name

**Comment by andrzejwaw** [comment]
> Two small things and 👍

fixed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 19]
```diff
@@ -54,6 +68,13 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
     }
 }
 
+fun Value?.toStatusCodeFilter(
+    accessLogFilterFactory: AccessLogFilterFactory
+): AccessLogFilterSettings.StatusCodeFilterSettings? {
+    val value = if (this?.stringValue != null) this.stringValue.toUpperCase() else return null
```
Comment (chemicL) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:74]: Wow, I didn't know you can "else return null".
This returns from function, not from if, right? Please simplify the control flow according to Kotlin idiomatic style. I'm sure this is easier if you use `?` properly.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt [hunk 20]
```diff
@@ -4,12 +4,15 @@ import com.google.protobuf.Struct
...
 import io.envoyproxy.controlplane.cache.NodeGroup
 import io.envoyproxy.envoy.api.v2.core.Node
+
 import pl.allegro.tech.servicemesh.envoycontrol.logger
 import pl.allegro.tech.servicemesh.envoycontrol.snapshot.SnapshotProperties
 
 class MetadataNodeGroup(val properties: SnapshotProperties) : NodeGroup<Group> {
     private val logger by logger()
 
+    private val accessLogFilterFactory = AccessLogFilterFactory()
```
Comment (chemicL) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt:14]: DI?

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt [hunk 21]
```diff
@@ -76,6 +78,8 @@ class EnvoyListenersFactory(
                     )
             )
 
+    private val accessLogFilterFactory = AccessLogFilterFactory()
```
Comment (chemicL) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt:81]: DI?

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+            )
+        }
+        val split = value.split(delimiter)
+        return AccessLogFilterSettings.StatusCodeFilterSettings(
+            comparisonOperator = ComparisonFilter.Op.valueOf(split[0]),
+            comparisonCode = split[1].toInt()
+        )
+    }
+
+    fun buildAccessLogStatusCodeFilter(builder: Builder, settings: AccessLogFilterSettings.StatusCodeFilterSettings) {
```
Comment (chemicL) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt:32]: How about an extension function on the builder instead of 2 params? Perhaps this would work?

    fun AccessLog.Builder.buildFromSettings(settings: AccessLogFilterSettings.StatusCodeFilterSettings)

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt [hunk 23]
```diff
@@ -0,0 +1,51 @@
+package pl.allegro.tech.servicemesh.envoycontrol.groups
```
[the remark this answers is not in the document]
Comment (chemicL) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt:1]: Actually, because it has two responsibilities:
- parsing
- producing envoy specific config

I'd split into two classes. The parsing one can stay here I guess. Or be moved to NodeMetadata class' file.

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroupTest.kt [hunk 24]
```diff
@@ -218,6 +225,82 @@ class MetadataNodeGroupTest {
...
+
+        // when
+        val group = nodeGroup.hash(Node.newBuilder().setMetadata(metadata?.build()).build())
+
+        // then
+        assertThat(group.listenersConfig!!.accessLogFilterSettings!!.statusCodeFilterSettings).isNull()
+    }
+
+    @Test
+    fun `should throw exception for access log filter status code empty value`() {
```
[the remark this answers is not in the document]
Comment (andrzejwaw) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroupTest.kt:280]: added testing empty input to NodeMetadataTest

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt (hunk 21 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt:81]: fixed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt (hunk 19 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:74]: fixed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt (hunk 20 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt:14]: fixed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/Groups.kt [hunk 25]
```diff
@@ -32,6 +32,7 @@ data class ListenersConfig(
     val accessLogPath: String = defaultAccessLogPath,
     val resourcesDir: String = defaultResourcesDir,
     val addUpstreamExternalAddressHeader: Boolean = defaultAddUpstreamExternalAddressHeader,
+    val accessLogFilterSettings: AccessLogFilterSettings? = null,
```
[the remark this answers is not in the document]
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/Groups.kt:35]: changed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt (hunk 22 above)
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt:32]: done

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt (hunk 23 above)
[the remark this answers is not in the document]
Comment (andrzejwaw) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/AccessLogFilterFactory.kt:1]: moved to another scope and splited functionality

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt [hunk 26]
```diff
@@ -318,4 +333,43 @@ class NodeMetadataTest {
...
     }
+
+    @ParameterizedTest
+    @MethodSource("validStatusCodeFilterData")
+    fun `should set statusCodeFilter for accessLogFilter`(input: String, op: ComparisonFilter.Op, code: Int) {
+        // given
+        val proto = accessLogFilterProto(statusCodeFilter = input)
+
+        // when
+        val incoming = proto.structValue?.fieldsMap?.get("status_code_filter").toStatusCodeFilter(
```
[the remark this answers is not in the document]
Comment (andrzejwaw) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt:344]: fixed

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt [hunk 27]
```diff
@@ -318,4 +333,43 @@ class NodeMetadataTest {
...
+
+    @ParameterizedTest
+    @CsvSource(
+        "LT:123",
+        "equal:400",
+        "eq:24",
+        "GT:200",
+        "testeq:400test"
+    )
+    fun `should not set statusCodeFilter for accessLogFilter for invalid status code filter data`(input: String) {
```
[the remark this answers is not in the document]
Comment (andrzejwaw) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataTest.kt:361]: fixed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 28]
```diff
@@ -71,8 +72,10 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
 fun Value?.toStatusCodeFilter(
     accessLogFilterFactory: AccessLogFilterFactory
 ): AccessLogFilterSettings.StatusCodeFilterSettings? {
-    val value = if (this?.stringValue != null) this.stringValue.toUpperCase() else return null
-    return accessLogFilterFactory.parseStatusCodeFilter(value)
+    return when (this?.stringValue) {
```
Comment (chemicL) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:75]: I see you replaced `if` with `when`.
The Kotlin idiomatic way to implement this:

```
fun Value?.toStatusCodeFilter(accessLogFilterFactory: AccessLogFilterFactory)
    : AccessLogFilterSettings.StatusCodeFilterSettings? = this?.stringValue?.let {
    accessLogFilterFactory.parseStatusCodeFilter(it.toUpperCase())
```

Check https://kotlinlang.org/docs/reference/null-safety.html#safe-calls

Comment (chemicL) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:75]: Also, there's no test case that validates null.
